haystack/eval.py

```diff
@@ -294,177 +341,246 @@ def calculate_f1_str_multi(gold_labels, prediction):
+#                                "doc_id": pred["label"].multiple_document_ids[i]}
+#                               for i in range(len(pred["label"].multiple_answers))]  # type: ignore
+#                 predicted_span = {"offset_start": answer["offset_start_in_doc"],
+#                                   "offset_end": answer["offset_end_in_doc"],
+#                                   "doc_id": answer["document_id"]}
+#
+#                 best_f1_in_gold_spans = 0
+#                 for gold_span in gold_spans:
+#                     if gold_span["doc_id"] == predicted_span["doc_id"]:
+#                         # check if overlap between gold answer and predicted answer
+#                         if not found_answer:
+#                             metric_counts, found_answer = _count_overlap(
+#                                 gold_span, predicted_span, metric_counts, answer_idx
+#                             )
+#                         # check for exact match
+#                         if not found_em:
+#                             metric_counts, found_em = _count_exact_match(
+#                                 gold_span, predicted_span, metric_counts, answer_idx
+#                             )
+#                         # calculate f1
+#                         current_f1 = _calculate_f1(gold_span, predicted_span)
+#                         if current_f1 > best_f1_in_gold_spans:
+#                             best_f1_in_gold_spans = current_f1
+#                 # top-1 f1
+#                 if answer_idx == 0:
+#                     metric_counts["summed_f1_top1"] += best_f1_in_gold_spans
+#                     metric_counts["summed_f1_top1_has_answer"] += best_f1_in_gold_spans
+#                 if best_f1_in_gold_spans > best_f1:
+#                     best_f1 = best_f1_in_gold_spans
+#
+#             if found_em:
+#                 break
+#
+#         # top-k answers: use best f1-score
+#         metric_counts["summed_f1_topk"] += best_f1
+#         metric_counts["summed_f1_topk_has_answer"] += best_f1
+#
+#     # question not answerable
+#     else:
+#         metric_counts["number_of_no_answer"] += 1
+#         metric_counts = _count_no_answer(pred["answers"], metric_counts)
+#
+#     return metric_counts
+
+
+def semantic_answer_similarity(predictions: List[List[str]],
+                               gold_labels: List[List[str]],
+                               sas_model_name_or_path: str = "sentence-transformers/paraphrase-multilingual-mpnet-base-v2") -> (np.float32, np.float32):
+    """
+    Computes Transformer-based similarity of predicted answer to gold labels to derive a more meaningful metric than EM or F1.
+    Returns per QA pair a) the similarity of the most likely prediction (top 1) to all available gold labels
+                        b) the highest similarity of all predictions to gold labels
```

Comment:
We are currently returning to float32 values here and not the individual predictions. Either we need to adjust the docstring or the returns. There is also a second sentence further down in the docstring that refers to this

Reply:
good point, the sas function is designed to work on batches, so it should return a sas score for each prediction. I changed the code accordingly.